### PR TITLE
make sdf_mutate friendlier and more predictable

### PR DESCRIPTION
**Pull request by mattpollock**
This resolves issue #1329

### CODE REVIEW

**Comment by mattpollock** [comment]
The documented example works without any chnges
```r
data(beavers, package = "datasets")
beaver_tbl <- copy_to(sc, beaver1, "beaver")
beaver_tbl %>%
  mutate(squared = temp ^ 2) %>%
  sdf_mutate(warm = ft_binarizer(squared, 1000)) %>%
  sdf_register("mutated")
#
#  # Source:   table<mutated> [?? x 6]
# # Database: spark_connection
#      day  time  temp activ squared  warm
#    <dbl> <dbl> <dbl> <dbl>   <dbl> <dbl>
#  1   346   840  36.3     0    1320  1.00
#  2   346   850  36.3     0    1321  1.00
#  3   346   900  36.4     0    1321  1.00
#  4   346   910  36.4     0    1326  1.00
#  5   346   920  36.6     0    1336  1.00
#  6   346   930  36.7     0    1346  1.00
#  7   346   940  36.7     0    1348  1.00
#  8   346   950  36.8     0    1351  1.00
#  9   346  1000  36.8     0    1355  1.00
# 10   346  1010  36.9     0    1360  1.00
# # ... with more rows
```

File: R/sdf_interface.R [hunk 1]
```diff
@@ -271,16 +271,13 @@ sdf_mutate_ <- function(.data, ..., .dots) {
...
       lazy_expr[[1]],        # function
-      data,                  # data
+      data,                  # data,
       output_col=output_col  # output column
-    )
+    ), inputs)               # inputs
 
-    call <- as.call(c(
-      preamble,
-      unlist(inputs)         # parsed (and evaluated if applicable), inputs
```
Comment (mattpollock) [inline R/sdf_interface.R]: In the `beavers` example 1000 was getting coerced to "1000" via `unlist` since the other list item was a column name string.

**Comment by mattpollock** [comment]
> The job exceeded the maximum time limit for jobs, and has been terminated.

Not sure what to do about that.

**Comment by kevinykuo** [comment]
@mattpollock Thanks for the PR! Looking at the issue you opened I'm thinking we can consider deprecating `sdf_mutate()` in favor of something like https://github.com/rstudio/sparklyr/issues/1267. Is your use case for `sdf_mutate()` basically to avoid typing a bunch of `"`s? I think we want to avoid having too many ways to achieve the same thing.

**Comment by mattpollock** [comment]
I see your point. I like using the mutate construct where I can name the output column on the LHS and forego the quotes. That makes spark API operations feel more like they belong in a dplyr pipeline. 

I feel pretty strongly that `sdf_mutate` should not stay as is given the odd behavior I pointed out in #1329, however if the direction you all are going in is to promote direct use `ft_*` and similar functions in pipelines I can deal with that. Would #1267 be a breaking change requiring all the current SE functions to be transitioned to NSE and (presumably) new `[ft|ml]_*._$` functions to be written to support standard evaluation?

**Comment by mattpollock** [comment]
I've been mulling this for a 1/2 hour now. I don't presume to have much of a vote here, but keeping NSE in verb functions seems more consistent. Let `ft_` and `ml_` use standard evaluation (i.e., don't do #1267) in the same way that `grepl`, `gsub`, etc. use standard evaluation in local `dplyr` pipelines. Locally most functions use SE and `dplyr` verbs do NSE. Following this pattern seems less confusing with the one caveat being that the `output_col=` argument is unique in the spark context. 

I'd _like_ to keep `sdf_mutate` around with the purposed modifications, but as I said above, if you want to kill it I can modify how I'm doing stuff today.

**Comment by kevinykuo** [comment]
Just gave this some thought, too. The plan for #1267 was that for `ft_.tbl_spark()` functions we'd support looking up unquoted names in the dataframe columns, then look for bindings in the calling environment, so it wouldn't break current usage. However, one of the workflows is that the user would interactively use these `ft_` and `ml_` functions with a data frame, e.g.
```
beaver_tbl %>%
  ft_binarizer("temp", "hot", 36.5) %>%
  ft_r_formula(hot ~ time + activ) %>%
  ml_logistic_regression()
```
then easily turn it into a pipeline by replacing the data frame input:
```
pipeline <- ml_pipeline(sc) %>%
  ft_binarizer("temp", "hot", 36.5) %>%
  ft_r_formula(hot ~ time + activ) %>%
  ml_logistic_regression()
```

NSE would complicate things here since it won't be obvious where we're looking for the variables first. This would be an argument for not implementing #1267 even if it's backwards compatible. (Also, it adds yet another way to do something...)

Perhaps we should patch `sdf_mutate()` now so it supports named arguments and you're unblocked. Also we could designate `sdf_mutate()` as the one place where NSE is supported for feature transformers. We want to move away from lazyeval though and use tidyeval instead.

@kevinushey @edgararuiz @javierluraschi thoughts here?

**Comment by kevinykuo** [comment]
OK we caught up a bit and this is what we'll do. We'll soft deprecate `sdf_mutate()`, so instead of

```
beaver_tbl %>%
  mutate(squared = temp ^ 2) %>%
  sdf_mutate(warm = ft_binarizer(squared, 1000))
```

we'd write

```
beaver_tbl %>%
  mutate(squared = temp ^ 2) %>%
  ft_binarizer("squared", "warm", 1000)
```

We can still use feature transformers without breaking pipelines. The `ft_` functions aren't SQL so this syntax makes that a bit more explicit. Also, the `sdf_` functions should really map 1:1 to Spark DataFrame functions. We can revisit implementing something in the future but it'd be a different name (or package.)

As for https://github.com/rstudio/sparklyr/issues/1267 we'll put it on hold for now. It'd be pretty involved and if we don't do it right it has the potential to cause much confusion and unexpected behavior.

Thanks again for the issue and your thoughts, @mattpollock !

**Comment by mattpollock** [comment]
Works for me. I'll adjust accordingly. Thanks very much.

**Pull request closed without merging**